### PR TITLE
feat: extract (epoch from col)

### LINKED ISSUE #2785: EXTRACT(EPOCH from column)
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

Datafusion supports a few date parts like "year" or "seconds" in the `EXTRACT` operator.

Postgres offers a few more. In particular it offers an `EPOCH` date part which extracts the number of seconds since 1970-01-01 00:00:00-00 (can be negative). 

See https://www.postgresql.org/docs/8.1/functions-datetime.html

My particular use case is using pre-existing grafana postgres plugin, which crafts queries with such a construct. But I think it would be generally useful to support converting a timestamp to a unix epoch.

**Describe the solution you'd like**

```sql
select EXTRACT(EPOCH from column) from mytable;
```

**Describe alternatives you've considered**

```sql
select CAST(column AS bigint)/1000000000 from mytable;
```

### PR DESCRIPTION
**Pull request by Weijun-H**
# Which issue does this PR close?

Closes #2785

# Rationale for this change
Explain #2785

# What changes are included in this PR?
Add `EXTRACT( EPOCH FROM COL)`

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sql/expr.rs [hunk 1]
```diff
@@ -1313,6 +1313,23 @@ async fn test_extract_date_part() -> Result<()> {
     Ok(())
 }
 
+#[tokio::test]
+async fn test_extract_epoch() -> Result<()> {
+    // test_expression!(
```
Comment (comphead) [inline datafusion/core/tests/sql/expr.rs:1318]: Thanks @Weijun-H are this commented tests needed?

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 2]
```diff
@@ -537,6 +539,35 @@ where
...
+                        TimeUnit::Microsecond => 1_000_000,
+                        TimeUnit::Nanosecond => 1_000_000_000,
+                    };
+
+                    let n: i64 = array.value(i).into();
+                    b.append_value(n as f64 / scale as f64);
+                }
+            }
+        }
+        _ => return Err(DataFusionError::Internal("Invalid data type".to_string())),
```
Comment (comphead) [inline datafusion/physical-expr/src/datetime_expressions.rs:566]: it might be useful to output the wrong datatype in error message

**Review by comphead — COMMENTED** [review]
Please add more tests to cover null values and different scales

File: datafusion/physical-expr/src/datetime_expressions.rs [hunk 3]
```diff
@@ -537,6 +539,40 @@ where
...
+    i64: From<T::Native>,
+{
+    let mut b = Float64Builder::with_capacity(array.len());
+    match array.data_type() {
+        DataType::Timestamp(tu, _) => {
+            for i in 0..array.len() {
+                if array.is_null(i) {
+                    b.append_null();
+                } else {
+                    let scale = match tu {
```
Comment (alamb) [inline datafusion/physical-expr/src/datetime_expressions.rs:554]: I double checked that this is consistent with postgres docs:  https://www.postgresql.org/docs/8.1/functions-datetime.html. 👍

**Review by alamb — APPROVED** [review]
Thanks @Weijun-H  -- I double checked the answers against postgres and it looks great. 


```sql
postgres=# select extract(epoch from '1870-01-01T07:29:10.256'::timestamp);
      extract
--------------------
 -3155646649.744000
(1 row)

postgres=# select extract(epoch from '2000-01-01T00:00:00.000'::timestamp);
     extract
------------------
 946684800.000000
(1 row)


postgres=# select extract(epoch from null::timestamp);
 extract
---------

(1 row)
```

**Comment by alamb** [comment]
Thank you @comphead  for the review

**Pull request merged**